Add `AhoCorasick`

Prabhat-Kumar-42 opened the pull request:
<!-- For completed items, change [ ] to [x] -->

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.

Prabhat-Kumar-42 commented:
Hey,
I added code for the Aho-Corasick algorithm.
This code should work for all the test cases.
Please review it, if any issue arises, please let me know.
I hope It will help the community.
Thank You !! :)

Manish-Rawat-1186 commented:
bro how did you convert into clang format

Prabhat-Kumar-42 commented:
> bro how did you convert into clang format
            
Hey, below are the steps for clang.
You can format your code with the command (run it somewhere _inside_ this project directory):
```shell
clang-format -i --style=file path/to/your/file.java
```
If you do not have the correct version of `clang-format` (i.e. < 16.0), you can remove this line:
https://github.com/TheAlgorithms/Java/blob/ea0eef128db5b3e5e1f9d58cb8a7e009cd508ead/.clang-format#L86
and add the _new line symbol_ at the end of your file manually (but do not commit the changed `.clang-format` file).

If you use gitpod, everything should run smoothly there.

_Originally posted by @vil02 in https://github.com/TheAlgorithms/Java/issues/4438#issuecomment-1741399556_

Prabhat-Kumar-42 commented:
I request the maintainer(s) to please merge my request using "hacktoberfest-accepted" label.
Thank You !!

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,187 @@
+        return countOfWords;
+    }
+
+    public static void main(String[] args) {
+        String[] patterns = {"ACC", "ATC", "CAT", "GCG", "C", "T"};
+        String text = "GCATCG";
+
+        AhoCorasick obj = new AhoCorasick();
+        obj.buildTrie(patterns);
+        obj.buildSuffixAndOutputLinks();
+
+        ArrayList<ArrayList<Integer>> res = obj.search(text);
+        ArrayList<Integer> countOfWords = obj.getRepeatCountOfWords();
+
+        System.out.println("Using Zero Based Indexing");
+        System.out.println("Dictonary is : ");
+        for (int i = 0; i < patterns.length; i++) {
+            System.out.println(i + ". " + patterns[i]);
+        }
+        System.out.println();
+        System.out.println("Given text is : " + text);
+        System.out.println();
+        System.out.println("-1 represents word is not in the given string");
+        for (int i = 0; i < patterns.length; i++) {
+            System.out.print(patterns[i] + " appeared " + countOfWords.get(i) + " times at indices : ");
+            if (res.get(i).isEmpty()) {
+                System.out.print(-1 + " ");
+            } else {
+                for (int endpoint : res.get(i)) {
+                    System.out.print((endpoint - patterns[i].length() + 1) + " ");
+                }
+            }
+            System.out.println();
+        }
+    }
```
Please remove the `main` method and instead add some proper tests. Good examples to look at:
- [`Factorial`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/main/java/com/thealgorithms/maths/Factorial.java) and [`FactorialTest`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/test/java/com/thealgorithms/maths/FactorialTest.java),
- [`FindMax`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/main/java/com/thealgorithms/maths/FindMax.java) and [`FindMaxTest`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/test/java/com/thealgorithms/maths/FindMaxTest.java).

```suggestion
```

Prabhat-Kumar-42 replied:
Yes, I understand.
Thank You for your review @vil02. 
I am working on adding test-file, I will get back to you as soon as I get it done.

Prabhat-Kumar-42 replied:
> Please remove the `main` method and instead add some proper tests. Good examples to look at:
> 
> * [`Factorial`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/main/java/com/thealgorithms/maths/Factorial.java) and [`FactorialTest`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/test/java/com/thealgorithms/maths/FactorialTest.java),
> * [`FindMax`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/main/java/com/thealgorithms/maths/FindMax.java) and [`FindMaxTest`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/test/java/com/thealgorithms/maths/FindMaxTest.java).

Hey @vil02,

I have resolved the issues you have mentioned, that is :

- Added a test-file for the code.
- Removed the main method.

Please review the code. If there's anything else that needs to be resolved regarding this, please do let me know.

Thank you!!

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,181 @@
+package com.thealgorithms.strings;
+
+import java.util.*;
```
```suggestion
import java.util.HashMap;
import java.util.ArrayList;
import java.util.Queue;
import java.util.LinkedList;
```

Prabhat-Kumar-42 replied:
Got it. Will do it.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,181 @@
+    private String[] patterns; // patterns according to which Trie is constructed
+
+    // Clears the Aho-Corasick data structures
+    public void clear() {
+        root = null;
+        if (res != null) {
+            res.clear();
+        }
+    }
```
This not not covered by any tests.

Prabhat-Kumar-42 replied:
Got it. Adding a test case for it.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,181 @@
+        }
+    }
+
+    Node root = null; // The root node of the Aho-Corasick trie
```
```suggestion
    private Node root = null; // The root node of the Aho-Corasick trie
```

Prabhat-Kumar-42 replied:
It's also done.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,181 @@
+                Node parentSuffix = currentState.suffix_link; // Get the parent's suffix link
+
+                // Calculate the suffix link for the child based on parent's suffix link
+                while (!parentSuffix.child.containsKey(cc) && parentSuffix != root) {
```
Tests do not cover all of the branches.

Prabhat-Kumar-42 replied:
Got it. Will add a test for this.

Prabhat-Kumar-42 replied:
Added test-case, regarding this with few more test-cases.
It is also done now.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,181 @@
+                if (parentSuffix.child.containsKey(cc)) {
+                    currentChild.suffix_link = parentSuffix.child.get(cc);
+                } else {
+                    currentChild.suffix_link = root;
```
This is not covered by any of the tests.

Prabhat-Kumar-42 replied:
Got it. Will add a test regarding this.

Prabhat-Kumar-42 replied:
Added test-case, regarding this with few more test-cases.
It is also done now.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,181 @@
+                while (parent != root && !parent.child.containsKey(ch)) {
+                    parent = parent.suffix_link;
+                }
+                if (parent.child.containsKey(ch)) {
```
In the existing tests, this condition if never `false`.

Prabhat-Kumar-42 replied:
Got it. Will add a test case regarding this.

Prabhat-Kumar-42 replied:
Added test-case, regarding this with few more test-cases.
It is also done now.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,181 @@
+        root.suffix_link = root; // Initialize the suffix link of the root to itself
+        Queue<Node> q = new LinkedList<>(); // Initialize a queue for BFS traversal
+
+        // Initialize suffix links for child nodes of the root
+        for (char rc : root.child.keySet()) {
+            q.add(root.child.get(rc)); // Add child node to the queue
+            root.child.get(rc).suffix_link = root; // Set suffix link to the root
+        }
```
This could be moved to a separate method (called something like `initializeSuffixLinksForChildNodesOfTheRoot`). In general, a good method name is the best comment.

Prabhat-Kumar-42 replied:
I understand. Will Implement this.

Prabhat-Kumar-42 replied:
Added this method. It's also done now.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,181 @@
+    }
+
+    Node root = null; // The root node of the Aho-Corasick trie
+    private ArrayList<ArrayList<Integer>> res; // Stores the positions where patterns are found in the text
```
What do you think about storing the result in another data structure, for example:
```suggestion
    private HashMap<String, ArrayList<Integer>> res;
```
so res['somePattern'] would be a list of positions where `somePattern` occurs.

Prabhat-Kumar-42 replied:
> What do you think about storing the result in another data structure, for example:
> 
> so res['somePattern'] would be a list of positions where `somePattern` occurs.

Hey @vil02 ,

I can do that, but it will cost more space and time.

The current algorithm uses indices to calculate positions, as it is more efficient in terms of calculation and comparison within the algorithm. If I implement it using strings, the trie data structure will need space according to the length of the string that needs to be saved in it. Additionally, I would still need to find a way to calculate the position of words, which in the above case can be solved using the stored integer.

**TL;DR - 1**: I can do it, but it will cost both time and space.

However, what I can do to somewhat mitigate this is to add a new `Hash<String, ArrayList<Integer>>` called `positionByString`. This data structure will use the stored data in `res` and the provided dictionary or set of patterns (in our case, the pattern array provided in the code) to store the combined result. This will keep the time complexity the same and will only require extra time to process `res` and the pattern array to populate `positionByString`.

**TL;DR - 2:** While still keeping `res`, can create a `Hash<String, ArrayList<Integer>>` which utilizes data from `res` and the `pattern array` to mitigate the overhead.

I would greatly appreciate hearing your thoughts on this.

Please feel free to share your insights and opinions.

Prabhat-Kumar-42 replied:
> What do you think about storing the result in another data structure, for example:
> 
> so res['somePattern'] would be a list of positions where `somePattern` occurs.

I ended up going with the **TL;DR - 2**, I have implemented the added functionality along with the test cases.

I would say it is done, but I would appreciate your feedback regarding this.

If you feel like any changes should be made, please do tell me.

vil02 commented on src/test/java/com/thealgorithms/strings/AhoCorasickTest.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Test
+    void testSearch() {
+        // Note - Indexing is zero based
+
+        // Test that the first pattern "ACC" is not found in the text
+        assertEquals(0, res.get(0).size());
+
+        // Test that the second pattern "ATC" is found at index 2
+        assertEquals(2, res.get(1).get(0));
+
+        // Test that the third pattern "CAT" is found at index 1
+        assertEquals(1, res.get(2).get(0));
+
+        // Test that the fourth pattern "GCG" is not found in the text
+        assertEquals(0, res.get(3).size());
+
+        // Test that the fifth pattern "C" is found at indices 1 and 4
+        assertEquals(2, res.get(4).size());
+        assertEquals(1, res.get(4).get(0));
+        assertEquals(4, res.get(4).get(1));
+
+        // Test that the sixth pattern "T" is found at index 3
+        assertEquals(3, res.get(5).get(0));
```
What do you think about more _direct test_ i.e. defining a variable `expected` and just checking if `res` and `expected` are equal.

Prabhat-Kumar-42 replied:
Got it.
Added direct test check style in test-files.
Please review it.

vil02 reviewed (changes requested):
Interesting contribution!

The code is quite complex - it could be simplified by introducing some new _helper_ methods. It is not easy, but we will make it perfect!

Prabhat-Kumar-42 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,181 @@
+
+    // Trie Node Class
+    class Node {
+        HashMap<Character, Node> child = new HashMap<>();
+        Node suffix_link;
+        Node output_link;
+        int pattern_ind;
```
I understand, will change the fields accordingly.

Prabhat-Kumar-42 commented:
> Interesting contribution!
> 
> The code is quite complex - it could be simplified by introducing some new _helper_ methods. It is not easy, but we will make it perfect!

Thank You @vil02 , and yes I agree with code being complex. People often face problem understanding AhoCorasick.
And Yes We Will make it perfect.   

I have made the changes you mentioned, and I've also added some more test cases along with the one you mentioned. Please review it.

If there's any issue regarding the code, please do tell me. I will fix it as soon as possible.

Thank you.

vil02 commented:
> And Yes We Will make it perfect.

That is the spirit!

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,253 @@
+    private String[] patterns; // patterns according to which Trie is constructed
+    private Queue<Node> q; // Used for BFS traversal
+
+    private Map<String, ArrayList<Integer>> positionByString; // store positions by string pattern
```
This is a bit redundant: it stores the same information as `res`.

vil02 reviewed (changes requested):
The `AhoCorasick` is quite big and I would suggest to cut it a bit.

What do you think about introducing a class called `Trie`. It should have:
- only one member `root`. It should have one constructor taking the `patterns`,
- one method `searchIn` taking the `text`.
The constructor would do the same as `buildTrie` and `buildSuffixAndOutputLinks`. The `searchIn` would return the data, which is currently stored in `res`.

The mentioned methods are the only ones, which need to `public`. Some methods like `clear()` are not needed.

For time being, everything can happen within `AhoCorasick`. One could think about giving it a `static` method `Map<String, ArrayList<Integer>> search(String text, String[] patterns);`. And a `private static` method doing the conversion from convert `ArrayList<ArrayList<Integer>>` to `Map<String, ArrayList<Integer>>`.

Prabhat-Kumar-42 commented:
> The `AhoCorasick` is quite big and I would suggest to cut it a bit.
> 
> What do you think about introducing a class called `Trie`. It should have:
> 
> * only one member `root`. It should have one constructor taking the `patterns`,
> * one method `searchIn` taking the `text`.
>   The constructor would do the same as `buildTrie` and `buildSuffixAndOutputLinks`. The `searchIn` would return the data, which is currently stored in `res`.
> 
> The mentioned methods are the only ones, which need to `public`. Some methods like `clear()` are not needed.
> 
> For time being, everything can happen within `AhoCorasick`. One could think about giving it a `static` method `Map<String, ArrayList<Integer>> search(String text, String[] patterns);`. And a `private static` method doing the conversion from convert `ArrayList<ArrayList<Integer>>` to `Map<String, ArrayList<Integer>>`.

Hey @vil02,

Since you asked to make modifications to the code and make new classes, i.e, Trie and AhoCorasick, I would like some suggestions on how should I structure these classes so It's organised and  benefits the Project. I have few options I have provided below.

### Option 1: Separate Files

Should I structure the code like this:
```
thealgorithms/
│   └── strings/
│       ├── AhoCorasick.java
│       ├── AhoCorasickTrie.java
│       ├── AhoCorasickTrieNode.java
│       └── OtherClasses/
└── OtherPackages/
```
In this option, `AhoCorasickTrieNode`, `AhoCorasickTrie`, and `AhoCorasick` have their separate classes in separate files, which helps with modularity and code manageability. `AhoCorasickTrieNode` implements the node of `AhoCorasickTrie`, `AhoCorasickTrie` implements the features you mentioned earlier, and `AhoCorasick` implements the static methods you mentioned.


### Option 2: Utility Package

Or should I structure it like this:
```
thealgorithms/
│   └── strings/
│             ├── AhoCorasick.java
│             ├── AhoCorasickUtilities/
│             │       ├── AhoCorasickTrie.java
│             │       └── AhoCorasickTrieNode.java
│             └── OtherClasses/
└── OtherPackages/
```
In this option, everything remains the same as in Option 1, but `AhoCorasickTrie.java` and `AhoCorasickTrieNode.java` are stored in the `AhoCorasickUtilities` package for ease of access and separation.

### Option 3: Inner Classes

Or should I make `Node` and `Trie` inner classes of the `AhoCorasick` class to keep the code in the same place?

Please advise on the best approach for code organization and project structure. Thank you!

vil02 commented:
@Prabhat-Kumar-42 I think that in case of this repository the **Option 3: Inner Classes** is the way to go.

The other suggestions are of curse the generally preferred ways, but they do not fit into this projects structure (usually there is _one algorithm - one file_ approach; there are exceptions, e.g. [`MedianOfRunningArray.java`](https://github.com/TheAlgorithms/Java/blob/master/src/main/java/com/thealgorithms/misc/MedianOfRunningArray.java), [`MedianOfRunningArrayDouble.java`](https://github.com/TheAlgorithms/Java/blob/master/src/main/java/com/thealgorithms/misc/MedianOfRunningArrayDouble.java), [`MedianOfRunningArrayFloat.java`](https://github.com/TheAlgorithms/Java/blob/master/src/main/java/com/thealgorithms/misc/MedianOfRunningArrayFloat.java)). A good thing is that the resulting file will not be that long.

Prabhat-Kumar-42 commented:
> @Prabhat-Kumar-42 I think that in case of this repository the **Option 3: Inner Classes** is the way to go.
> 
> The other suggestions are of curse the generally preferred ways, but they do not fit into this projects structure (usually there is _one algorithm - one file_ approach; there are exceptions, e.g. [`MedianOfRunningArray.java`](https://github.com/TheAlgorithms/Java/blob/master/src/main/java/com/thealgorithms/misc/MedianOfRunningArray.java), [`MedianOfRunningArrayDouble.java`](https://github.com/TheAlgorithms/Java/blob/master/src/main/java/com/thealgorithms/misc/MedianOfRunningArrayDouble.java), [`MedianOfRunningArrayFloat.java`](https://github.com/TheAlgorithms/Java/blob/master/src/main/java/com/thealgorithms/misc/MedianOfRunningArrayFloat.java)). A good thing is that the resulting file will not be that long.

I appreciate your feedback. I will Implement **Option 3: Inner Classes**.

Also, I have one more thing that I need your take on - 

 - since you mentioned keeping some members private (e.g. - `private static method doing the conversion from convert ArrayList<ArrayList<Integer>> to Map<String, ArrayList<Integer>>`). It won't be possible to test them in the test file as they can't be accessed. 

 - should I make these functions public or just test the available public methods and members of the code? 

Thank You !!

Prabhat-Kumar-42 commented:
> The `AhoCorasick` is quite big and I would suggest to cut it a bit.
> 
> What do you think about introducing a class called `Trie`. It should have:
> 
> * only one member `root`. It should have one constructor taking the `patterns`,
> * one method `searchIn` taking the `text`.
>   The constructor would do the same as `buildTrie` and `buildSuffixAndOutputLinks`. The `searchIn` would return the data, which is currently stored in `res`.
> 
> The mentioned methods are the only ones, which need to `public`. Some methods like `clear()` are not needed.
> 
> For time being, everything can happen within `AhoCorasick`. One could think about giving it a `static` method `Map<String, ArrayList<Integer>> search(String text, String[] patterns);`. And a `private static` method doing the conversion from convert `ArrayList<ArrayList<Integer>>` to `Map<String, ArrayList<Integer>>`.



> @Prabhat-Kumar-42 I think that in case of this repository the **Option 3: Inner Classes** is the way to go.


Hey @vil02 ,

I have implemented the changes you mentioned. I have kept the File Structure as **Option 3: Inner Classes**:

- Added Class Trie:
  - It has only one method, `searchIn(text)`, returning data stored in `res`.
  - The constructor would do the same as `buildTrie` and `buildSuffixAndOutputLinks`.

- AhoCorasick:
  - It has static methods: `Map<String, ArrayList<Integer>> search(String text, String[] patterns)` and `private static Map<String, ArrayList<Integer>> convert(ArrayList<ArrayList<Integer>>)`.

- Removed unimportant methods:
  - `clear()`.

- Updated Test Cases According to the updated AhoCorasick Source File.

There's one thing to note: Instead of having only a single member `root`, I have added `patterns` and `res` as members too, so methods don't look too complicated with lengthy parameters passing through them.

Please review it.

If there's any issue regarding the code, please do tell me. I will fix it as soon as possible.

Thank you.

vil02 commented on src/test/java/com/thealgorithms/strings/AhoCorasickTest.java:
```diff
@@ -0,0 +1,138 @@
+        text = "GCATCG";
+    }
+
+    @Test
+    void testSearch() {
+        // Test searching for multiple patterns in the input text.
+        Map<String, ArrayList<Integer>> positions = AhoCorasick.search(text, patterns);
+
+        // Define expected results for each pattern
+        int[][] expected = {
+            {}, // "ACC" should not be found
+            {2}, // "ATC" should be found at index 2
+            {1}, // "CAT" should be found at index 1
+            {}, // "GCG" should not be found
+            {1, 4}, // "C" should be found at indices 1 and 4
+            {3} // "T" should be found at index 3
+        };
+
+        // Check specific pattern occurrences
+        for (int i = 0; i < patterns.length; i++) {
+            assertIterableEquals(toList(expected[i]), positions.get(patterns[i]));
+        }
+    }
```
Please notice that this test could still pass, if the `search` method would return a map with some additional keys.

What do you think about the following:
```suggestion
    @Test
    void testSearch() {
        final var expected = Map.of("ACC", new ArrayList<>(Arrays.asList()), "ATC", new ArrayList<>(Arrays.asList(2)), "CAT", new ArrayList<>(Arrays.asList(1)), "GCG", new ArrayList<>(Arrays.asList()), "C", new ArrayList<>(Arrays.asList(1, 4)), "T", new ArrayList<>(Arrays.asList(3)));

        assertEquals(expected, AhoCorasick.search(text, patterns));
    }
```

In case of normal line lengths, the code above looks really nice (in my opinion).

Make sure to
```java
import java.util.Arrays;
```

Please apply above idea in other tests.

Prabhat-Kumar-42 replied:
yes, got it, will implement it.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,228 @@
+/*********************************************************************************
+ *                                                                               *
+ *          Aho-Corasick String Matching Algorithm Implementation                *
+ *                                                                               *
+ * This code implements the Aho-Corasick algorithm, which is used for efficient  *
+ * string matching in a given text. It can find multiple patterns simultaneously *
+ *                 and records their positions in the text.                      *
+ *                                                                               *
+ *                      Author: Prabhat-Kumar-42                                 *
+ *               GitHub: https://github.com/Prabhat-Kumar-42                     *
+ *                                                                               *
+ *********************************************************************************/
```
I would suggest to document the code in more _doxygen_-like style.

Prabhat-Kumar-42 replied:
sure, got it, will do it.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,228 @@
+         *      list[n] contains the list of start-index of word pattern[n]
+         */
+
+        private ArrayList<ArrayList<Integer>> res; // Stores positions where patterns are found in the text
```
Why is this a member? I think it should be a local variable in `searchIn`.

Prabhat-Kumar-42 replied:
I made it a member because : 

   - I need it to allocate list of all the patterns[] members positions and to do that I have to fist assign it's size as patterns[] which I did in constructor, but I also have to assign all it's i-th position i.e. res[i] = empty ArrayList<>(). If i do all that in patterns I have to I have to run the loop ( which is equal to patterns.length ) two times (2n times if n is size of patterns)

       - 1st time for building `Trie`

       -  2nd time for this empty ArrayList<>() allocation to `res` indices in `searchIn`.

  - If I let it be member : 

       - I only have to run the loop for a single time, i.e. when building the tree in constructor the object, empty ArrayList<>() will be assigned to all res indices simultaneously.

       - I don't have to pass lengthy parameter to setUpStartPoints().

***TL;DR : it was just to save time by running the loop a single time instead of twice.***

Although Since you mentioned I should make it local variable in 'searchIn' . I will implement that, If there's any Issue, please do tell me, I will address it asap.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,228 @@
+         */
+
+        private ArrayList<ArrayList<Integer>> res; // Stores positions where patterns are found in the text
+        private String[] patterns; // patterns according to which Trie is constructed
```
```suggestion
        private final String[] patterns; // patterns according to which Trie is constructed

```

Prabhat-Kumar-42 replied:
yes, got it, will do that.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,228 @@
+            root.setSuffixLink(root); // Initialize the suffix link of the root to itself
+            Queue<Node> q = new LinkedList<>(); // Initialize a queue for BFS traversal
+
+            // Initialize suffix links for child nodes of the root
+            for (char rc : root.getChild().keySet()) {
+                Node childNode = root.getChild().get(rc);
+                q.add(childNode); // Add child node to the queue
+                childNode.setSuffixLink(root); // Set suffix link to the root
+            }
+            while (!q.isEmpty()) {
+                Node currentState = q.poll(); // Get the current node for processing
+
+                // Iterate through child nodes of the current node
+                for (char cc : currentState.getChild().keySet()) {
+                    Node currentChild = currentState.getChild().get(cc); // Get the child node
+                    Node parentSuffix = currentState.getSuffixLink(); // Get the parent's suffix link
+
+                    // Calculate the suffix link for the child based on the parent's suffix link
+                    while (!parentSuffix.getChild().containsKey(cc) && parentSuffix != root) {
+                        parentSuffix = parentSuffix.getSuffixLink();
+                    }
+
+                    // Set the calculated suffix link or default to root
+                    if (parentSuffix.getChild().containsKey(cc)) {
+                        currentChild.setSuffixLink(parentSuffix.getChild().get(cc));
+                    } else {
+                        currentChild.setSuffixLink(root);
+                    }
+
+                    q.add(currentChild); // Add the child node to the queue for further processing
+                }
+
+                // Establish output links for nodes to efficiently identify patterns within patterns
+                if (currentState.getSuffixLink().getPatternInd() >= 0) {
+                    currentState.setOutputLink(currentState.getSuffixLink());
+                } else {
+                    currentState.setOutputLink(currentState.getSuffixLink().getOutputLink());
+                }
+            }
+        }
```
This should be cut into smaller pieces.

Prabhat-Kumar-42 replied:
> * one method `searchIn` taking the `text`.
>   The constructor would do the same as `buildTrie` and `buildSuffixAndOutputLinks`. The `searchIn` would return the data, which is currently stored in `res`.

Ah yes, It might be my misunderstanding because when I read the above I thought you need only one function in  `Trie class` that is `searchIn` and  let `constructor` do the work of the `buildTrie` and `buildSuffixAndOutputLinks` and merged them into constructor.

I will revert them back, if there's still any issue please let me know.

vil02 replied:
Constructors can also call other methods.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,228 @@
+                }
+            }
+
+            /*
+             * By default res contains end-points.
+             * This function converts those endpoints to start points
+             */
+            for (int i = 0; i < patterns.length; i++) {
+                if (!res.get(i).isEmpty()) {
+                    for (int j = 0; j < res.get(i).size(); j++) {
+                        int endpoint = res.get(i).get(j);
+                        res.get(i).set(j, endpoint - patterns[i].length() + 1);
+                    }
+                }
+            }
```
This could go to a separate `private` method.

Prabhat-Kumar-42 replied:
same reason as above for why i did it, but no problem, I will revert it back into separate method.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,228 @@
+        return convert(res, patterns);
+    }
+
+    // Private static method for converting results to a map
```
We all see that is is `private static`

Prabhat-Kumar-42 replied:
got it. will remove it.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,228 @@
+        }
+    }
+
+    // Public static method to search for patterns in text
```
It is written line below that it is `public` and `static`.

Prabhat-Kumar-42 replied:
got it, will remove it.

Prabhat-Kumar-42 commented:
Hey @vil02 ,

I have updated the issues you have mentioned.

Please review the code.

Thank You !!

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,236 @@
+        private Node root = null; // Root node of the trie
+        private final String[] patterns; // patterns according to which Trie is constructed
+
+        public Node getTrieRoot() {
+            return root;
+        }
```
Is is nowhere used. Is there a reason to have this method? If yes: please write tests for it.

```suggestion
```

Prabhat-Kumar-42 replied:
got it, removed it.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,236 @@
+            return suffixLink;
+        }
+
+        public void setSuffixLink(Node suffixLink) {
```
```suggestion
        public void setSuffixLink(final Node suffixLink) {

```

Prabhat-Kumar-42 replied:
yes, added it.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,236 @@
+            return outputLink;
+        }
+
+        public void setOutputLink(Node outputLink) {
```
```suggestion
        public void setOutputLink(final Node outputLink) {

```

Prabhat-Kumar-42 replied:
yes, added it.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,236 @@
+            return patternInd;
+        }
+
+        public void setPatternInd(int patternInd) {
```
```suggestion
        public void setPatternInd(final int patternInd) {

```

Prabhat-Kumar-42 replied:
yes, added it.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,236 @@
+            return root;
+        }
+
+        public Trie(String[] patterns) {
```
```suggestion
        public Trie(final String[] patterns) {

```

Prabhat-Kumar-42 replied:
yes, added it.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,236 @@
+            }
+        }
+
+        // Builds the suffix links and output links in the trie
```
The method name just says that.

```suggestion
```

Prabhat-Kumar-42 replied:
yes, got it, done

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,236 @@
+            root.setSuffixLink(root); // Initialize the suffix link of the root to itself
+            Queue<Node> q = new LinkedList<>(); // Initialize a queue for BFS traversal
+
+            // Initialize suffix links for child nodes of the root
```
```suggestion
```

Prabhat-Kumar-42 replied:
done

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,236 @@
+        }
+        // by default res contains end-points. This function converts those
+        // endpoints to start points
+        private void setUpStartPoints(ArrayList<ArrayList<Integer>> res) {
```
`res` is not the perfect name here.

Prabhat-Kumar-42 replied:
got it. renamed res to positionByStringIndexValue

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,236 @@
+        // endpoints to start points
+        private void setUpStartPoints(ArrayList<ArrayList<Integer>> res) {
+            for (int i = 0; i < patterns.length; i++) {
+                if (!res.get(i).isEmpty()) {
+                    for (int j = 0; j < res.get(i).size(); j++) {
+                        int endpoint = res.get(i).get(j);
+                        res.get(i).set(j, endpoint - patterns[i].length() + 1);
+                    }
+                }
```
I think this `if` statement is not needed here. If it would be `false`, then the loop has length 0 and nothing happens anyway.
```suggestion
                for (int j = 0; j < res.get(i).size(); j++) {
                    int endpoint = res.get(i).get(j);
                    res.get(i).set(j, endpoint - patterns[i].length() + 1);
                }
```

Prabhat-Kumar-42 replied:
done.

vil02 commented on src/test/java/com/thealgorithms/strings/AhoCorasickTest.java:
```diff
@@ -0,0 +1,164 @@
+        // Define the expected results for each pattern
+        final var expected = Map.of("ACC", new ArrayList<>(Arrays.asList()), "ATC", new ArrayList<>(Arrays.asList(2)), "CAT", new ArrayList<>(Arrays.asList(1)), "GCG", new ArrayList<>(Arrays.asList()), "C", new ArrayList<>(Arrays.asList(1, 4)), "T", new ArrayList<>(Arrays.asList(3)));
+
+        // Perform the search using the Aho-Corasick algorithm
+        final var positions = AhoCorasick.search(text, patterns);
+
+        // Compare the actual results with the expected results
+        assertEquals(expected, positions);
```
There is no need for these comments.
```suggestion
        assertEquals(expected, AhoCorasick.search(text, patterns));
```

Prabhat-Kumar-42 replied:
got it, implemented it.

vil02 commented on src/test/java/com/thealgorithms/strings/AhoCorasickTest.java:
```diff
@@ -0,0 +1,164 @@
+        final var searchPatterns = new String[] {"CG", "TCG", "ATCG"};
+
+        // Define the expected positions (4, 3, 2) for the patterns
+        final var expected = new int[] {4, 3, 2};
```
Why not to define `expected` as `Map` as in other tests?

Prabhat-Kumar-42 replied:
got it, implemented it.

vil02 commented on src/test/java/com/thealgorithms/strings/AhoCorasickTest.java:
```diff
@@ -0,0 +1,164 @@
+        final var positions = AhoCorasick.search(text, searchPatterns);
+
+        // Check that the positions contain all the provided patterns in a case-insensitive way
+        for (String pattern : searchPatterns) {
+            assertTrue(positions.containsKey(pattern));
+        }
```
Why not to define `expected` and check it it the same as `positions`?

Prabhat-Kumar-42 replied:
got it, implemented it.

Prabhat-Kumar-42 commented:
Hey @vil02 ,

I have updated the changes you have mentioned.

Please review the code.

If there's anything that needs to be modified, please let me know.

Thank You !!

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,227 @@
+import java.util.Map;
+import java.util.Queue;
+
+public class AhoCorasick {
```
```suggestion
public final class AhoCorasick {
    private AhoCorasick() {
    }
```

Prabhat-Kumar-42 replied:
got it. will add the final field.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,227 @@
+public class AhoCorasick {
+
+    // Trie Node Class
+    public class Node {
```
```suggestion
    private static class Node {

```

Prabhat-Kumar-42 replied:
got it, will change it to private static.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,227 @@
+    }
+
+    // Trie Class
+    public class Trie {
```
```suggestion
    public static class Trie {

```

Prabhat-Kumar-42 replied:
got it. added static to Trie class.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,227 @@
+        }
+
+        // Searches for patterns in the input text and records their positions
+        public ArrayList<ArrayList<Integer>> searchIn(String text) {
```
```suggestion
        public ArrayList<ArrayList<Integer>> searchIn(final String text) {

```

Prabhat-Kumar-42 replied:
got it. added the final field.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,227 @@
+    }
+
+    // method to search for patterns in text
+    public static Map<String, ArrayList<Integer>> search(String text, String[] patterns) {
```
```suggestion
    public static Map<String, ArrayList<Integer>> search(final String text, final String[] patterns) {

```

Prabhat-Kumar-42 replied:
got it. added the final field.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,227 @@
+    }
+
+    // method for converting results to a map
+    private static Map<String, ArrayList<Integer>> convert(ArrayList<ArrayList<Integer>> positionByStringIndexValue, String[] patterns) {
```
```suggestion
    private static Map<String, ArrayList<Integer>> convert(final ArrayList<ArrayList<Integer>> positionByStringIndexValue, final String[] patterns) {

```

Prabhat-Kumar-42 replied:
got it. added the final field.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,227 @@
+
+    // method to search for patterns in text
+    public static Map<String, ArrayList<Integer>> search(String text, String[] patterns) {
+        Trie trie = new AhoCorasick().new Trie(patterns);
```
This is a very strange construction. Making the _nested_ classes `static`, will allow to do that:
```suggestion
        final var trie = new Trie(patterns);

```

Prabhat-Kumar-42 replied:
implemented this.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,227 @@
+    // method to search for patterns in text
+    public static Map<String, ArrayList<Integer>> search(String text, String[] patterns) {
+        Trie trie = new AhoCorasick().new Trie(patterns);
+        ArrayList<ArrayList<Integer>> positionByStringIndexValue = trie.searchIn(text);
```
```suggestion
        final var positionByStringIndexValue = trie.searchIn(text);
```

Prabhat-Kumar-42 replied:
implemented this

vil02 commented on src/test/java/com/thealgorithms/strings/AhoCorasickTest.java:
```diff
@@ -0,0 +1,119 @@
+
+package com.thealgorithms.strings;
+
+import static org.junit.jupiter.api.Assertions.*;
```
```suggestion
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.assertEquals;
```

Prabhat-Kumar-42 replied:
got it. now changed it to this.

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,227 @@
+             *      ......
+             *      list[n] contains the list of start-index of word pattern[n]
+             */
+            ArrayList<ArrayList<Integer>> positionByStringIndexValue = new ArrayList<>(patterns.length); // Stores positions where patterns are found in the text
+            Node parent = root; // Start searching from the root node
+
+            for (int i = 0; i < patterns.length; i++) {
+                positionByStringIndexValue.add(new ArrayList<Integer>()); // Initialize a list to store positions of the current pattern
+            }
```
```suggestion
            var positionByStringIndexValue = initializePositionByStringIndexValue(patterns.length); // Stores positions where patterns are found in the text
            Node parent = root; // Start searching from the root node
```

Prabhat-Kumar-42 replied:
got it, implemented and used initializePositionByStringIndexValue().

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,227 @@
+                    parent = parent.getChild().get(ch); // Update the current node to the child node
+
+                    // If the current node represents a pattern, record its position in positionByStringIndexValue
+                    if (parent.getPatternInd() > -1) {
+                        positionByStringIndexValue.get(parent.getPatternInd()).add(i);
+                    }
+
+                    Node outputLink = parent.getOutputLink();
+                    // Follow output links to find and record positions of other patterns
+                    while (outputLink != null) {
+                        positionByStringIndexValue.get(outputLink.getPatternInd()).add(i);
+                        outputLink = outputLink.getOutputLink();
+                    }
```
This ask to be in a separate method. In order to make it _nice_ I would suggest to create another class like `RawSearchResult`. It could be either derived from `ArrayList<ArrayList<Integer>>` (I would call it a _tricky_ approach) or simply a class having one member of the type `ArrayList<ArrayList<Integer>>`.

This comment is not a _must_.

Prabhat-Kumar-42 replied:
Hey @vil02 ,
I tried implementing what you said above. I used this
 
> simply a class having one member of the type `ArrayList<ArrayList<Integer>>`.

I hope you like it.
 
Still if there's any thing that needs to be added or if you want me to revert it back to how it was, please do tell me. I will address it asap.

Prabhat-Kumar-42 commented:
Hey @vil02 ,

I have updated the changes you have mentioned.

I tried what you said above about wrapping up pattern indexing in another class. I hope you like it.

Please review the code.

If there's anything that needs to be modified, please let me know.

Thank You !!

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,247 @@
+import java.util.Queue;
+
+public final class AhoCorasick {
+
```
```suggestion
    private AhoCorasick() {
    }

```

vil02 commented on src/main/java/com/thealgorithms/strings/AhoCorasick.java:
```diff
@@ -0,0 +1,247 @@
+        private ArrayList<ArrayList<Integer>> positionByStringIndexValue;
+
+        // Constructor to initialize the recorder with the position list
+        public PatternPositionRecorder(final ArrayList<ArrayList<Integer>> positionByStringIndexValue) {
```
```suggestion
        PatternPositionRecorder(final ArrayList<ArrayList<Integer>> positionByStringIndexValue) {
```

vil02 commented:
@Prabhat-Kumar-42 wow, this was a pull-request. Thanks!

vil02 commented:
Closes #4464.

Prabhat-Kumar-42 commented:
Thanks @vil02 , It was fun working with you and I learned a lot in the process.
I am looking forward to contributing more and working with you.
Thank you again !!
